### DISCUSSION TITLE
curl-agent/curl-add idea

### DISCUSSION THREAD

**Discussion by icing** [curl/curl#13954] [Ideas]
This is an idea about using a resident *agent* to hold information where `curl` can store/retrieve information that is relevant for transfers. This should similar to `ssh-agent` (or maybe even use that, more below) for `ssh` clients.

The agent would *not* make any transfers itself, nor have anything to do with connections. Instead it could hold (some of) the following information for clients with `urls` being the lookup key:
- credentials
- TLS sessions
- HTTPS records
- IPv4/6, https: eyeballing outcomes

Which of these are a good idea remains to be decided. There are security implications and usability improvements which need careful analysis..

### How it would work

On startup, a connection to an agent is attempted. Any failure to do so would fallback to current behaviour. Talking to the agent is purely optional. When curl detects it needs any information, it would retrieve that from the agent. On obtaining new items, it would try to store them in the agent (however, it would never add credentials on its own). 

Parallel curl instances have no guarantee of the order in which items are stored. There is no synchronisation between concurrent runs. Each piece of information needs to stand on its own.

As with `ssh`, the agent would allow connecting only on a unix domain socket and only for the same user.

### Credentials

The use case should be quite obvious: instead of adding `--user name:passw` to the command line, one issues *once* a command like `curl-add credentials https://somedomain --user name:passw` and all curl commands can use it. 

This eliminates several security issues with mentioning credentials on the command line/in scripts or passing them around as environment variables. However it would mean that *any* user process is able to look this up as well. This is very different from `ssh` where the keys stored in the agent or not exposed to the client. The agent instead signs requests or even gets signatures from a hardware device.

But for `https:` and many other protocols, user names and passwords are required more or less in the clear. An exception is http `digest` authentication, but deployment of that scheme on the web seems minor. However **if** credentials are passed around or placed in the environment or a `.netrc` file, having them in an agent process seems more secure.

One could also imagine `curl-add` looking up credentials in a password manager or an external device which can then be removed once the loading is done.

### TLS sessions

TLS sessions could be shared via an agent, so after a `curl` transfer, the TLS session is stored and available for the next run. This may be useful in faster connection handshakes and TLS early data. There needs to be a limit on the number of sessions stored in an agent, though.

### HTTPS records, ipv4/6/https eyeballing

Just performance related. Save another DNS lookup or start a connection with the better IP or HTTP version first.

## using ssh-agent

The protocol that is talked between an `ssh-agent` and clients is [published here](https://datatracker.ietf.org/doc/html/draft-miller-ssh-agent#constraints). The interesting part is that the *extension* mechanism for new record types. This *might* be usable for curl.

It seem worth getting feedback from the OpenSSH maintainers on the suitability of such an approach. The benefits would be great as the agent implementation requires robustness, memory protection, concurrency handling and many other things which are already there.

## Summary

An agent process for curl could enhance security, usability and performance for people who want it. It should not intrude on the "standalone" workings for curl.

**Comment by djmdjm** [comment]
My conceptual advice would be to think hard about what you're trying to achieve with a putative `curl-agent`, because what you exclude from the concept is at least as important than what you include. `ssh-agent` is a successful design because fulfils a usability goal at the same time as meeting security goals: users benefit from the convenience of only needing to type their private keys passphrases once per session (usability), while their keys are stored in a relatively small TCB that can implement protections are difficult to realise in a full standalone application (security).

Another mental model for ssh-agent that has been useful at various times is that it's a cheapskate's version of a HSM - a user can put sensitive stuff into it (keys), and make use of the sensitive stuff in it (signature operations) but they can't get the sensitive stuff out of it, and we try to make it hard for an attacker too.

Practically, if you agree with either of these conceptions, then it suggests stating small and moving incrementally. I'm not a TLS expert but a safe store of TLS session tickets and/or mTLS client keys seems like a reasonable place to start.

As for the ssh-agent protocol itself: it's basically what we inherited from SSH1 and it's fine. If you ever intend to support SSH objects in your agent or if you intend to support proxying to an actual ssh-agent for some operations then maybe it makes sense to be compatible, but otherwise IMO the protocol doesn't matter as much as the security philosophy you're using.

**Comment by bagder** [comment]
Where do we draw the line for what data/information the agent should keep? A few of the caches might be helpful to have there too, depending on how we think it is to be used: alt-svc and HSTS data for example.

**Reply by icing** [reply]
I think we cannot say where we'll end up. `curl-agent` should remain an optional thing. We may, at least initially, not have it for all platforms. Also, there needs to be a limit on the amount of data we keep or we'll end up implementing a database for domain name things - which does already seem to exist on the internet.

**Comment by Spiffyk** [comment]
Hi, I wanted to follow up from the Curl Up discussion today with a maybe slightly alternative idea about the proposed daemon. Keep in mind I'm not (yet) familiar with curl's internals so the idea may not make sense with the current architecture of curl.

With regards to connection sharing, what if the daemon only took care of the data transport and connection management? I.e. the curl daemon would be the one establishing connections to servers and actually sending/receiving data (maybe sharing send/recv buffers via shm/memfd/non-Linux equivalent with the client to avoid extra copies), while the curl client would be doing all the rest of what curl normally does, possibly also utilizing the separate curl-agent for secrets.

In my mind this could let curl boost the performance with connection sharing while avoiding problems with unwanted state retained between curl invocations, like the current working directory.
